packages/smooth_app/lib/pages/product/nutrition_page_loaded.dart

```diff
@@ -651,68 +653,26 @@ class _NutritionPageLoadedState extends State<NutritionPageLoaded> {
   NutritionUnit _detectUnit(Product product) {
     if (product.nutriments == null) {
       return NutritionUnit.per100g;
-    } else if (<dynamic>[
-          product.nutriments!.saturatedFatServing,
-          product.nutriments!.proteinsServing,
-          product.nutriments!.novaGroupServing,
-          product.nutriments!.energyServing,
-          product.nutriments!.carbohydratesServing,
-          product.nutriments!.caffeineServing,
-          product.nutriments!.calciumServing,
-          product.nutriments!.ironServing,
-          product.nutriments!.vitaminAServing,
-          product.nutriments!.vitaminB1Serving,
-          product.nutriments!.vitaminB2Serving,
-          product.nutriments!.vitaminB6Serving,
-          product.nutriments!.vitaminB9Serving,
-          product.nutriments!.vitaminB12Serving,
-          product.nutriments!.vitaminCServing,
-          product.nutriments!.vitaminDServing,
-          product.nutriments!.vitaminEServing,
-          product.nutriments!.vitaminKServing,
-          product.nutriments!.vitaminPPServing,
-          product.nutriments!.magnesiumServing,
-          product.nutriments!.phosphorusServing,
-          product.nutriments!.potassiumServing,
-          product.nutriments!.sodiumServing,
-          product.nutriments!.zincServing,
-          product.nutriments!.copperServing,
-          product.nutriments!.seleniumServing,
-          product.nutriments!.cholesterolServing,
-          product.nutriments!.butyricAcidServing,
-          product.nutriments!.caproicAcidServing,
-          product.nutriments!.caprylicAcidServing,
-          product.nutriments!.lauricAcidServing,
-          product.nutriments!.myristicAcidServing,
-          product.nutriments!.palmiticAcidServing,
-          product.nutriments!.stearicAcidServing,
-          product.nutriments!.oleicAcidServing,
-          product.nutriments!.linoleicAcidServing,
-          product.nutriments!.docosahexaenoicAcidServing,
-          product.nutriments!.eicosapentaenoicAcidServing,
-          product.nutriments!.erucicAcidServing,
-          product.nutriments!.monounsaturatedServing,
-          product.nutriments!.polyunsaturatedServing,
-          product.nutriments!.alcoholServing,
-          product.nutriments!.pantothenicAcidServing,
-          product.nutriments!.biotinServing,
-          product.nutriments!.chlorideServing,
-          product.nutriments!.chromiumServing,
-          product.nutriments!.fluorideServing,
-          product.nutriments!.iodineServing,
-          product.nutriments!.manganeseServing,
-          product.nutriments!.molybdenumServing,
-          product.nutriments!.omega3FatServing,
-          product.nutriments!.omega6FatServing,
-          product.nutriments!.transFatServing
-        ].firstWhere((dynamic element) => element != null,
-            orElse: () => null) !=
-        null) {
+    } else if (_hasNutrientsInServingSize(product)) {
```

Comment:
I don't say your code is a bad translation, I say that what was coded before was wrong.

> "nutrition_data_per" field will be populated by either "100g" or "serving", if relevant

cf. https://github.com/openfoodfacts/openfoodfacts-dart/issues/565

Reply:
I can't follow here, what exactly is the problem (and how should I go about solving it)

Reply:
Would be something like
```dart
if (product.nutrimentDataPer == PerSize.serving.offTag) {
  return NutritionUnit.perServing;
}
return NutritionUnit.per100g;
```

It's `product.nutrimentDataPer` that tells if the data for "product as sold" is for 100g or serving.
It should also be set accordingly when we save the data to the server.

Reply:
Applied this 👍🏼